8340818: Add a new jtreg test root to test the generated documentation

nizarbenalla opened the pull request:
Please review this change that adds a new test root `docs` dedicated to testing the documentation, which has been a work in progress for a while. Testings for links, encoding, HTML, accessibility will be later added in following PRs. 

We also define a new make target `test-docs` meant for local use and depends on the docs.
This also adds the necessary configurations needed at Oracle.

TIA

erikj79 commented on test/docs/jdk/javadoc/TestDocs.java:
```diff
@@ -0,0 +1,84 @@
+public class TestDocs {
+    private static final Path ROOT_PATH = Path.of(System.getProperty("test.jdk"));
+
+    public static Path resolveDocs() {
+        Path firstCandidate = ROOT_PATH.getParent()
+                .resolve("docs");
+        Path secondCandidate = ROOT_PATH.getParent().getParent()
+                .resolve("docs.doc_api_spec").resolve("docs");
+
+        if (Files.exists(firstCandidate)) {
+            return firstCandidate;
+        } else if (Files.exists(secondCandidate)) {
+            return secondCandidate;
+        } else {
+            throw new SkippedException("docs folder not found in either location");
+        }
+    }
```
This shouldn't be needed. We should define an environment variable that is expected to point to the docs image that RunTests.gmk sets up. Perhaps there could be java code to automatically fallback to looking in the first candidate location here, as that's the layout of the build itself. Some people prefer to bypass the makefiles and run jtreg standalone, and that would help support that usecase. I would put that java code in a common library class for the whole test root.

nizarbenalla replied:
Understood
> Some people prefer to bypass the makefiles and run jtreg standalone, and that would help support that usecase

I tend to do that most of the time if I only made changes to the test.

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -1069,6 +1069,52 @@ var getJibProfilesProfiles = function (input, common, data) {
+
+    profiles = concatObjects(profiles, testOnlyProfilesPrebuiltDocs);
+
+    if (!new java.io.File(__DIR__, "../../README.md").exists()) {
```
There are several conditionals and blocks that are duplicated from the run-tests-prebuilt definitions above. I think I would prefer if you merged this into the originals instead of duplicating things. I think it's more likely that future changes will need to update all prebuilt profiles and that will be easier to get right without duplication.

nizarbenalla commented on make/conf/jib-profiles.js:
```diff
@@ -1069,6 +1069,52 @@ var getJibProfilesProfiles = function (input, common, data) {
+                "jcov", testedProfileJdk, testedProfileTest, "docs.doc_api_spec",
+            ],
+            make_args: testOnlyMake,
+            environment: {
```
Fixed in [8fcc977](https://github.com/openjdk/jdk/pull/21272/commits/8fcc9775fdabeedb3f14255bf44c1a1e080ab70a).

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -999,55 +999,37 @@ var getJibProfilesProfiles = function (input, common, data) {
         testOnlyProfilesPrebuilt["run-test-prebuilt"]["dependencies"].push(testedProfile + ".jdk_symbols");
     }
 
+    var testOnlyProfilesPrebuiltDocs = clone(testOnlyProfilesPrebuilt);
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"] = testOnlyProfilesPrebuiltDocs["run-test-prebuilt"];
+    delete testOnlyProfilesPrebuiltDocs["run-test-prebuilt"];
```
I think this could be simplified to something like (untested):
```suggestion
    var testOnlyProfilesPrebuiltDocs = {};
    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"] = clone(testOnlyProfilesPrebuiltDocs["run-test-prebuilt"]);
```

nizarbenalla replied:
This exact code didn't work, but I now have a more concise version

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -999,55 +999,37 @@ var getJibProfilesProfiles = function (input, common, data) {
-    // On macosx add the devkit bin dir to the path in all the run-test profiles.
-    // This gives us a guaranteed working version of lldb for the jtreg failure handler.
-    if (input.build_os == "macosx") {
-        macosxRunTestExtra = {
-            dependencies: [ "lldb" ],
-            environment_path: [
-                input.get("gnumake", "install_path") + "/bin",
-                input.get("lldb", "install_path") + "/Xcode/Contents/Developer/usr/bin",
-            ],
-        };
-        profiles["run-test"] = concatObjects(profiles["run-test"], macosxRunTestExtra);
-        profiles["run-test-prebuilt"] = concatObjects(profiles["run-test-prebuilt"], macosxRunTestExtra);
-    } else if (input.build_os == "windows") {
-        // On windows, add the devkit debugger to the path in all the run-test profiles
-        // to make them available to the jtreg failure handler.
-        var archDir = "x64";
-        if (input.build_arch == "aarch64") {
-            archDir = "arm64"
-        }
-        windowsRunTestExtra = {
-            environment_path: [
-                input.get("devkit", "install_path") + "/10/Debuggers/" + archDir
-            ]
+        if (profileToCheck != null) {
+            targetProfile[runTestProfile]["target_os"] = profileToCheck["target_os"];
+            targetProfile[runTestProfile]["target_cpu"] = profileToCheck["target_cpu"];
         }
-        profiles["run-test"] = concatObjects(profiles["run-test"], windowsRunTestExtra);
-        profiles["run-test-prebuilt"] = concatObjects(profiles["run-test-prebuilt"], windowsRunTestExtra);
```
What happened to all of this?

nizarbenalla replied:
I somehow accidentally removed this

magicus commented on test/docs/ProblemList.txt:
```diff
@@ -0,0 +1,10 @@
+#
+# Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+# ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
```
Please change this to the standard GPL header.

magicus commented on test/docs/TEST.ROOT:
```diff
@@ -0,0 +1,28 @@
+# This file identifies the root of the test-suite hierarchy.
```
This file should have a standard copyright header, too. I see now that the existing TEST.ROOT in jdk, langtools and jaxp is also missing this header. If you want, you can add it to them in this PR too, or we'll open a separate issue for that. But in any case this new file needs a proper header.

magicus commented on make/Main.gmk:
```diff
@@ -1334,6 +1340,7 @@ RUN_TEST_TARGETS := $(addprefix run-test-, $(ALL_NAMED_TESTS))
 
 run-test: test
 exploded-run-test: exploded-test
+run-test-docs: test-docs
```
`run-tests` is legacy support for a deprecated syntax. It should not be extended to new targets.

nizarbenalla commented:
Thanks for the reviews, I have one last patch I want to add to this PR. Will include copyright headers to all files in it.

magicus commented:
I'm wondering a bit on how the new `test-docs` target works with the pre-existing pattern of `test-<test>`, for the test names that we detect in FindTests.gmk. If anyone would add a jtreg group named `docs`, things would be very weird. 

Right now there is a `docs_all` group, which mean that you can run `make test-docs_all`. But this will not automatically bring in the dependency on the docs, and so will fail unless this has already been run.

It seems that the intention here is that you should always use `make test-docs TEST=docs_all` to run the tests in the new `docs` root.

I'm not sure if this is the best, or even the right, way to do this.

magicus commented:
Ideally, we would be able to analyse what jtregs are requested to run, and add the proper dependencies. This seems unfortunately hard to do in the general case, meaning that if you run like `make test TEST=some/specific/test/in/docs` we are unlikely to be able to defer that a dependency on the docs are needed.

But otoh, we could do so at least for `test-docs_all`. 

I'm thinking as I type right now, so bear with me. But I believe what we need is not a new target (`test-docs`), but a way to express that testing needs additional dependencies. Let's introduce a `TEST_DEPS` control variable instead! Then we can basically do:
```
$(eval $(call SetupTarget, test, \
    MAKEFILE := RunTests, \
    ARGS := TEST="$(TEST)", \
    DEPS := jdk-image test-image $(TEST_DEPS), \
))

$(eval $(call SetupTarget, exploded-test, \
    MAKEFILE := RunTests, \
    ARGS := TEST="$(TEST)" JDK_IMAGE_DIR=$(JDK_OUTPUTDIR), \
    DEPS := exploded-image test-image $(TEST_DEPS), \
))

```

Now the user will need to type `make test-docs_all TEST_DEPS=docs-jdk` (or `make test TEST=docs_all TEST_DEPS=docs-jdk`) to run the docs tests, which is flexible, but not very user friendly. So we should also need something along these lines:

```
test-docs: test-docs_all
```
to get a nice alias for testing docs.

(cont'd)

magicus commented:
Finally, we need to append `docs-jdk` to `TEST_DEPS` automatically in the cases we can determine that a test in the docs test root is requested. 

One way of doing this is to look for `test-docs_all` and `test-docs` in `$(MAKECMDGOALS)`, and look for `docs_all` in `$(TEST)`. Perhaps we can also extend that to look for `docs/...` in `$(TEST)`. 

That will end up in a bit of duplicated code, though. I'm thinking if there is a way to intercept this dependency much closer to the `SetupTarget` of `test` and `exploded-test`, which is where both `make test-docs_all` and `make test TEST=docs_all` end up. Currently, I'm a bit stuck here. I'll need to think a bit more on that.

magicus commented:
I think this is hard to solve fully generally, but I think we could add something along these lines:
```
$(eval $(call AddTestDependency, docs_all, docs-jdk))
```

and have a function in MainSupport in something along these lines: (warning! untested code)
```
define AddTestDependency
  test-$(strip $1): $2

  exploded-test-$(strip $1): $2

  ifeq ($(filter $(TEST), $1)
    TEST_DEPS += $2
  endif
endef
```

magicus commented:
I believe we have some old quirks and workarounds that we might be able to change into using `AddTestDependency` as a more general solution, but that is out of scope for this PR.

nizarbenalla commented:
Is it okay to assume that if someone wants to test the generated docs, he will have most luckily built them beforehand?
So even if users use `make test TEST=some/specific/test/in/docs`, it can work.

I added the new test target for convenience.

erikj79 commented:
> Is it okay to assume that if someone wants to test the generated docs, he will have most luckily built them beforehand?
> So it can work if users use `make test TEST=some/specific/test/in/docs`, while not being ideal.

I definitely think it's worth trying to figure out a reasonably well functioning way to automate this dependency as Magnus is trying to suggest. People in general expect make targets to have proper dependencies. If you change something and run `make test...` you expect the relevant part of the build to rerun before starting the tests. Having to remember to run multiple things in order leads to mistakes.

nizarbenalla commented on test/docs/TEST.ROOT:
```diff
@@ -0,0 +1,28 @@
+# should be taken to handle test failures of intermittent or
+# randomness tests.
+
+keys=intermittent randomness needs-src needs-src-jdk_javadoc
```
I removed all of these keys, I don't think any of them are necessary.

nizarbenalla commented:
I added the copyright header to `TEST.ROOT` files, and added a standard comment at the beginning of `docs/problemlist.txt`

I dropped the `test-docs` make target, until I can add a better way of passing the dependency.

nizarbenalla commented:
Sorry, one last push. Reverting changes to the other `TEST.ROOT` files until I can tell what the appropriate header for them is.

nizarbenalla commented:
I merged with master and saw a lot more files changed than expected, it was because the base branch from pr/21272 was outdated. I fixed it now.

magicus commented:
> I dropped the `test-docs` make target, until I can add a better way of passing the dependency.

Do you think you can implement the `TEST_DEPS` solution as sketched by me above, or do you need my help to do it?

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -999,25 +999,36 @@ var getJibProfilesProfiles = function (input, common, data) {
+
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].dependencies.push("docs.doc_api_spec", "tidy");
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["DOCS_IMAGE_DIR"] = input.get("docs.doc_api_spec", "install_path");
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["TIDY_HOME"] = input.get("tidy", "home_path") + "/bin/tidy";
```
This should be all you need for supplying `TIDY` to the makefiles when running tests without configure.
```suggestion
    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["TIDY"] = input.get("tidy", "home_path") + "/bin/tidy";
```

magicus commented on make/RunTestsPrebuiltSpec.gmk:
```diff
@@ -68,6 +68,9 @@ JAVA_FLAGS_SMALL := -XX:+UseSerialGC -Xms32M -Xmx512M -XX:TieredStopAtLevel=1
 BUILDJDK_JAVA_FLAGS_SMALL := -Xms32M -Xmx512M -XX:TieredStopAtLevel=1
 BUILD_JAVA_FLAGS := $(JAVA_FLAGS_BIG)
 
+# Express that testing needs additional dependencies
```
You don't need to set this. In fact, it makes it impossible to set it from the command line.

magicus commented:
There is a list of `INIT_CONTROL_VARIABLES` in `InitSupport.gmk`. You should add `TEST_DEPS` to it, to stop make from complaining about an unknown control variable.

Also adding to the record that I and Nizar spoke privately, and agreed that as a first step we only provide the `TEST_DEPS` hook, and save the part of setting it automatically depending on the target test as a follow-up work.

nizarbenalla commented:
With the latest patch, running  `make test TEST=docs_all TEST_DEPS=docs-jdk` works as intended but ` make test-docs_all TEST_DEPS=docs-jdk` does not.

I see that there is a function `DeclareRunTestRecipe` used to add prefixes, but all `test-$1` targets have the same two dependencies so it might be difficult to add custom dependencies when using that syntax

` $(foreach t, $(filter-out test-make%, $(ALL_TEST_TARGETS)), $(eval $t: jdk-image test-image))`

magicus commented:
At this point, it is simplest to just add the `$(TEST_DEPS)` dependency there:

```
  $(foreach t, $(filter-out test-make%, $(ALL_TEST_TARGETS)), $(eval $t: jdk-image test-image $(TEST_DEPS)))
  $(foreach t, $(ALL_EXPLODED_TEST_TARGETS), $(eval $t: exploded-image test-image $(TEST_DEPS)))
```

We will need to revisit this anyway if we want to calculate TEST_DEPS automatically for certain test targets.

magicus commented on test/docs/req.flg:
```diff
@@ -0,0 +1 @@
+echo_file test/TEST.properties
```
What is this for? I see that there is a similar file for the `jdk` test suite, but it says `echo_file test/TEST.ROOT`.

nizarbenalla replied:
I will admit that I only added it because I saw it in the `jdk` test suit, I wanted to keep the same structure.

Should I remove it?

erikj79 commented on make/RunTests.gmk:
```diff
@@ -868,6 +869,13 @@ define SetupRunJtregTestBody
 
   $1_JTREG_BASIC_OPTIONS += -e:TEST_IMAGE_DIR=$(TEST_IMAGE_DIR)
 
+  ifneq ($$(DOCS_IMAGE_DIR), )
+    DOCS_PATH := $$(DOCS_IMAGE_DIR)
+  else
+    DOCS_PATH := $$(OUTPUTDIR)/images/docs
+  endif
+  $1_JTREG_BASIC_OPTIONS += -e:DOCS_IMAGE_DIR=$$(DOCS_PATH)
```
Looking at this again, I just realized we do have this path defined in spec.gmk already, but with a different name: `DOCS_JDK_IMAGE_DIR` (along with `DOCS_JAVASE_IMAGE_DIR` and `DOCS_REFERENCE_IMAGE_DIR`). I think we should stick with this existing name instead of introducing a new one, so it's consistent across the whole build system. Then we don't need this fallback to $(OUTPUTDIR) anymore. `DOCS_JDK_IMAGE_DIR` will either be defined in spec.gmk from configure or from jib-profiles when running `run-tests-prebuilt-docs`.

nizarbenalla replied:
Thanks, I hadn't seen this

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -999,25 +999,36 @@ var getJibProfilesProfiles = function (input, common, data) {
+    };
+
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].dependencies.push("docs.doc_api_spec", "tidy");
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["DOCS_IMAGE_DIR"] = input.get("docs.doc_api_spec", "install_path");
```
Change to DOCS_JDK_IMAGE_DIR.

erikj79 commented on test/docs/tools/tester/DocTester.java:
```diff
@@ -0,0 +1,47 @@
+ * Test framework for performing tests on the generated documentation.
+ */
+public class DocTester {
+    private final static String DIR = System.getenv("DOCS_IMAGE_DIR");
```
Change to DOCS_JDK_IMAGE_DIR.

magicus commented:
@nizarbenalla Did you verify that `make test-docs_all TEST_DEPS=docs-jdk` works now?

nizarbenalla commented on make/RunTests.gmk:
```diff
@@ -869,12 +869,7 @@ define SetupRunJtregTestBody
-    DOCS_PATH := $$(OUTPUTDIR)/images/docs
-  endif
-  $1_JTREG_BASIC_OPTIONS += -e:DOCS_IMAGE_DIR=$$(DOCS_PATH)
+  $1_JTREG_BASIC_OPTIONS += -e:DOCS_JDK_IMAGE_DIR=$$(DOCS_JDK_IMAGE_DIR)
```
I didn't include a null check here as `DOCS_JDK_IMAGE_DIR` will always hold a value (e.g `build/.../images/docs`) even if the docs don't actually exist. I will be up to the tests to skip when docs aren't found.

erikj79 replied:
It's fine to leave out the conditional, but the double `$$` isn't needed and I'm surprised if it actually works. That must be some lucky double evaluation somewhere in that case. 

```suggestion
  $1_JTREG_BASIC_OPTIONS += -e:DOCS_JDK_IMAGE_DIR=$(DOCS_JDK_IMAGE_DIR)
```

magicus replied:
I'm not sure why you are saying that. This is in the body of `SetupRunJtregTest`, which will be eval:ed when called, as usual. So `$$` should be used for all variable expansions. It's actually the other way around; `$` will only work due to pure "luck" in this case, and we kind of mis-used that before, splitting the set of variables into two parts, one there `$` would suffice, and one where `$$` were needed. This was just confusing, to save an extra keypress, so I've tried to use `$$` everywhere and always in these macro bodies, for consistency.

erikj79 replied:
Oh, I misread the context. Magnus is correct.

nizarbenalla commented:
The last patch includes Eric's suggestions. I've ran multiple tests to make sure running `tidy` works, as well as testing the generated documentation. 

Expressing dependencies works fine whether you use `make test-$1` or `make test TEST=...`

erikj79 commented on make/conf/jib-profiles.js:
```diff
@@ -999,25 +999,36 @@ var getJibProfilesProfiles = function (input, common, data) {
+    };
+
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].dependencies.push("docs.doc_api_spec", "tidy");
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["DOCS_JDK_IMAGE_DIR"] = input.get("docs.doc_api_spec", "install_path");
+    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["TIDY"] = input.get("tidy", "home_path") + "/bin/tidy";
```
If we are doing another round here anyway, could you try to break up these lines a bit? I'm not enforcing strict 80 chars in the js file, but this is wider than a comfortable side-by-side diff is able to render for me.

```suggestion
    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["DOCS_JDK_IMAGE_DIR"]
            = input.get("docs.doc_api_spec", "install_path");
    testOnlyProfilesPrebuiltDocs["run-test-prebuilt-docs"].environment["TIDY"]
            = input.get("tidy", "home_path") + "/bin/tidy";
```

erikj79 commented:
> I'd like to avoid integrating this right before the weekend, and I still want to add copyright headers to the other TEST.ROOT files if I can find out when they were added. I roughly know how to do it.

I would say fixing copyrights in the other test roots is definitely out of scope for this change.

nizarbenalla commented:
I will need to file followup RFEs for things that were discussed during code review
